distributed/protocol/highlevelgraph.py

```diff
@@ -0,0 +1,115 @@
+from tlz import valmap
+
+from dask.core import keys_in_tasks
+from dask.highlevelgraph import HighLevelGraph, Layer
+
+from ..utils import (
+    str_graph,
+    tokey,
+    CancelledError,
+)
+
+from .utils import (
+    msgpack_opts,
+)
+from .serialize import (
+    import_allowed_module,
+    msgpack_encode_default,
+    msgpack_decode_default,
+)
+
+
+def _materialized_layer_pack(
+    layer: Layer, all_keys, known_key_dependencies, allowed_client, allows_futures
+):
+    from distributed.utils_comm import unpack_remotedata
+    from distributed.worker import dumps_task
+
+    dsk = {k: unpack_remotedata(v, byte_keys=True) for k, v in layer.items()}
+    unpacked_futures = set.union(*[v[1] for v in dsk.values()]) if dsk else set()
+    for future in unpacked_futures:
+        if future.client is not allowed_client:
+            raise ValueError(
+                "Inputs contain futures that were created by another client."
+            )
+        if tokey(future.key) not in allows_futures:
+            raise CancelledError(tokey(future.key))
+    unpacked_futures_deps = {}
+    for k, v in dsk.items():
+        if len(v[1]):
+            unpacked_futures_deps[k] = {f.key for f in v[1]}
+    dsk = {k: v[0] for k, v in dsk.items()}
+
+    missing_keys = set(dsk.keys()).difference(known_key_dependencies.keys())
+    dependencies = {
+        k: keys_in_tasks(all_keys, [dsk[k]], as_list=False) for k in missing_keys
+    }
+    for k, v in unpacked_futures_deps.items():
+        dependencies[k] = set(dependencies.get(k, ())) | v
+
+    # The scheduler expect all keys to be strings
+    dependencies = {
+        tokey(k): [tokey(dep) for dep in deps] for k, deps in dependencies.items()
+    }
+    dsk = str_graph(dsk, extra_values=all_keys)
+    dsk = valmap(dumps_task, dsk)
+    return {"dsk": dsk, "dependencies": dependencies}
+
+
+def highlevelgraph_pack(hlg: HighLevelGraph, allowed_client, allows_futures):
+    layers = []
+
+    # Dump each layer (in topological order)
+    for layer in (hlg.layers[name] for name in hlg._toposort_layers()):
+        if not layer.is_materialized():
+            state = layer.__dask_distributed_pack__()
+            if state is not None:
+                layers.append(
+                    {
+                        "__module__": layer.__module__,
+                        "__name__": type(layer).__name__,
+                        "state": state,
+                    }
+                )
+                continue
+
+        # Falling back to the default serialization, which will materialize the layer
+        layers.append(
+            {
+                "__module__": None,
+                "__name__": None,
+                "state": _materialized_layer_pack(
+                    layer,
+                    hlg.get_all_external_keys(),
+                    hlg.key_dependencies,
+                    allowed_client,
+                    allows_futures,
+                ),
+            }
+        )
```

Comment:
What if this was the implementation of `Layer.__dask_distributed_pack__`?  That might simplify things a bit here

Reply:
We would import the relevant `distributed` functions within the method in the Dask codebase.

Reply:
It's not a big deal either way, but it might isolate things a bit.

Reply:
I think it is best to have the fall back in Distributed for now. Long-term it would be nice to have it in `Layer.__dask_distributed_pack__()` but right now it requires a lot of very Distributed specific things like `key_dependencies`, `allowed_client`, and `allows_futures`. 
Furthermore, my plan is to also incorporate [`dsk.map_basic_layers()`](https://github.com/dask/distributed/blob/4d2dab8d9bf1dc8fa498c4ec95cc0edec1a36435/distributed/client.py#L2598) into `distributed_pack()` in order to remove both `map_basic_layers()` and `map_tasks()` from the HLG API.